src/System.Management.Automation/utils/ClrFacade.cs

```diff
@@ -58,9 +58,36 @@ internal static IEnumerable<Assembly> GetAssemblies(TypeResolutionState typeReso
         /// </param>
         internal static IEnumerable<Assembly> GetAssemblies(string namespaceQualifiedTypeName = null)
         {
-            return PSAssemblyLoadContext.GetAssembly(namespaceQualifiedTypeName) ??
-                   AssemblyLoadContext.Default.Assemblies.Where(a =>
-                       !a.FullName.StartsWith(TypeDefiner.DynamicClassAssemblyFullNamePrefix, StringComparison.Ordinal));
+            return PSAssemblyLoadContext.GetAssembly(namespaceQualifiedTypeName) ?? GetPSVisibleAssemblies();
+        }
+
+        /// <summary>
+        /// Return assemblies from the default load context and the 'individual' load contexts.
+        /// The 'individual' load contexts are the ones holding assemblies loaded via 'Assembly.Load(byte[])' and 'Assembly.LoadFile'.
+        /// Assemblies loaded in any custom load contexts are not consider visible to PowerShell to avoid type identity issues.
+        /// </summary>
+        private static IEnumerable<Assembly> GetPSVisibleAssemblies()
+        {
+            const string IndividualAssemblyLoadContext = "System.Runtime.Loader.IndividualAssemblyLoadContext";
```

Comment:
My concern is that `IndividualAssemblyLoadContext` is _internal_ class and .Net is free to change the name.
Maybe more reliable to use an assembly context name (`"Assembly.Load(byte[], ...)"` and `"Assembly.LoadFile({0})"`). It is public contract and has one prefix `Assembly.Load` for both.

Reply:
Any one can create a ALC instance with the name "Assembly.Load(byte[], ...)" or "Assembly.LoadFile(".
It's less reliable to depend on the name than the type. They can change the type name, but I think that possibility is very low, and we have test to catch the failure if the type name really changes.